patches/libtock-rs/07-debug_allocations.patch

```diff
@@ -0,0 +1,98 @@
+diff --git a/Cargo.toml b/Cargo.toml
+index 386a9ed..af3c5db 100644
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -10,6 +10,7 @@ linked_list_allocator = { version = "0.6.6", default-features = false }
+ 
+ [features]
+ panic_console = []
++debug_allocations = []
+ 
+ [dev-dependencies]
+ corepack = { version = "0.4.0", default-features = false, features = ["alloc"] }
+diff --git a/src/entry_point.rs b/src/entry_point.rs
+index 2fe5c40..c978ee5 100644
+--- a/src/entry_point.rs
++++ b/src/entry_point.rs
+@@ -368,22 +368,77 @@ pub unsafe extern "C" fn rust_start(app_start: usize, stacktop: usize, app_heap_
+     }
+ }
+ 
++#[cfg(feature = "debug_allocations")]
++use core::fmt::Write;
++#[cfg(feature = "debug_allocations")]
++use core::sync::atomic;
++#[cfg(feature = "debug_allocations")]
++use core::sync::atomic::AtomicUsize;
++
+ #[cfg(any(target_arch = "arm", target_arch = "riscv32"))]
+ #[global_allocator]
+-static ALLOCATOR: TockAllocator = TockAllocator;
++static ALLOCATOR: TockAllocator = TockAllocator::new();
+ 
+ static mut HEAP: Heap = Heap::empty();
+ 
+-struct TockAllocator;
++struct TockAllocator {
++    #[cfg(feature = "debug_allocations")]
++    count: AtomicUsize,
++    #[cfg(feature = "debug_allocations")]
++    size: AtomicUsize,
++}
++
++impl TockAllocator {
++    const fn new() -> TockAllocator {
++        TockAllocator {
++            #[cfg(feature = "debug_allocations")]
++            count: AtomicUsize::new(0),
++            #[cfg(feature = "debug_allocations")]
++            size: AtomicUsize::new(0),
++        }
++    }
++}
+ 
+ unsafe impl GlobalAlloc for TockAllocator {
+     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+-        HEAP.allocate_first_fit(layout)
++        let ptr = HEAP
++            .allocate_first_fit(layout)
+             .ok()
+-            .map_or(0 as *mut u8, |allocation| allocation.as_ptr())
++            .map_or(ptr::null_mut(), NonNull::as_ptr);
++        #[cfg(feature = "debug_allocations")]
++        {
++            self.count.fetch_add(1, atomic::Ordering::SeqCst);
++            self.size.fetch_add(layout.size(), atomic::Ordering::SeqCst);
++            writeln!(
++                crate::console::Console::new(),
++                "alloc[{}, {}] = {:?} ({} ptrs, {} bytes)",
++                layout.size(),
++                layout.align(),
++                ptr,
++                self.count.load(atomic::Ordering::SeqCst),
++                self.size.load(atomic::Ordering::SeqCst)
```

Comment:
I think it's preferable to use the result of `fetch_add` (and `fetch_sub` for deallocation below) and either:
- Recompute the new count and size locally.
- Just print the old values (which is equivalent anyway since we also print the size).

This way we avoid race conditions.

Reply:
Libtock apps are single-threaded and not preemptible, except at `yield` points where control-flow can jump to a callback. So there's no risk of race condition here and the displayed results are correct.

I used `AtomicUsize` because it has interior mutability while being `Sync`, therefore we can have a `static TockAllocator` object not marked as `mut` and avoid using `unsafe` (accessing `static mut` objects is unsafe in general, notably due to the possibility of multiple threads). If I recall correctly, that wouldn't be possible with a `Cell`, because the Rust compiler forbids `static` objects that are not `Sync` (due to the possibility of multi-threaded access, even though libtock isn't multi-threaded).

If libtock became multi-threaded in the future:
- Fetching the values once wouldn't remove the risk of a race condition between fetching `count` and `size`.
- Even though some printed values may be incorrect, the overall count would still converge to a correct value due to the use of atomics, in the sense that non-racy (de)allocations would print correct values.

Overall, I find it clearer to perform each arithmetic operation only once, rather than fetching fewer times. Given the trade-offs, it's mostly a matter of style, and the scope is limited to some debugging scenarios anyway.

Reply:
I see. Then I think it's worth documenting why `AtomicUsize` is used (next to `TockAllocator`).